austince/prometheus
# Add base xDS REST SD and kuma_sd implementation

Implementation of a [Kuma](kuma.io) Service Discovery mechanism via the [xDS REST protocol](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#rest-json-polling-subscriptions).

Since this is new code, the protobuf library used is the official `google.golang.com/protobuf` lib, which is where the community would like to migrate to, as per [this ML thread](https://groups.google.com/g/prometheus-developers/c/uFWRyqZaQis/m/Bi28HO4pAgAJ). The usage is entirely isolated from the `prompb` package where the `gogo/proto` package is currently used. When this transition is complete in Prometheus, it may make sense to include the Kuma MADS v1 `.proto` and compile it here, but since this is a "frozen" API it did not seem too bad to include the compiled version here in the meantime.

In upstream PR, will close:
https://github.com/prometheus/prometheus/issues/7919

## Testing

1. Download and unzip the testing files [kuma-sd-testing-files.zip](https://github.com/austince/prometheus/files/6423696/kuma-sd-testing-files.zip)
2. [Install Kuma](https://kuma.io/docs/1.1.2/installation/helm/) via helm and the `mads-kuma-values.yaml`, which uses images built from `kumahq/kuma:master@97ebe14a3`
2.1. `helm install --namespace kuma-system kuma kuma/kuma -f mads-kuma-values.yaml --create-namespace` 
3. Install the Kuma demo w/ `./mads-demo-install.sh`
4. Install Prometheus via the `mads-prom-metrics.yaml` which will use the `austince/prometheus:kuma-xds-sd` image
5. Generate some load w/ `./mads-demo-gen-load.sh`
6. Port forward Prometheus and see scraped metrics
6.1. `kubectl -n kuma-metrics port-forward svc/prometheus-server 9090:80`

### Screenshots

![Screen Shot 2021-05-04 at 4 42 09 PM](https://user-images.githubusercontent.com/4655775/117067395-f60ed980-acf7-11eb-9956-ab2e9d8fd080.png)
Target synchronization

![Screen Shot 2021-05-04 at 4 41 44 PM](https://user-images.githubusercontent.com/4655775/117067386-f3ac7f80-acf7-11eb-81cb-739e5f4ba488.png)
Targets in Prometheus

![Screen Shot 2021-05-04 at 4 43 21 PM](https://user-images.githubusercontent.com/4655775/117067398-f7400680-acf7-11eb-8d09-cfa127f8b2c5.png)
Scraped Envoy metrics
